src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java

```diff
@@ -201,6 +206,312 @@ public int hashCode() {
+                loc, String.format("slice start must be an integer, not '%s'", startObj));
+          }
+          if (endObj == Runtime.NONE) {
+            stop = step < 0 ? lower : upper;
+          } else if (endObj instanceof Integer) {
+            stop = (Integer) endObj;
+            if (stop < 0) {
+              stop += length;
+              stop = Math.max(stop, lower);
+            } else {
+              stop = Math.min(stop, upper);
+            }
+          } else {
+            throw new EvalException(loc, String.format("slice end must be an integer, not '%s'", endObj));
+          }
+          return new Slice(start, stop, step);
+        }
+      }
+
+      private final int start;
+      private final int stop;
+      private final int step;
+      private final int size;
+
+      private RangeListView(int start, int stop, int step) {
+        this.start = start;
+        this.stop = stop;
+        this.step = step;
+        this.size = computeSize(start, stop, step);
+      }
+
+      @Override
+      public Integer get(int index) {
+        int value = start + step * index;
+        if ((step > 0 && value > stop) || (step < 0 && value < stop)) {
+          throw new ArrayIndexOutOfBoundsException(index);
+        }
+        return value;
+      }
+
+      @Override
+      public int size() {
+        return size;
+      }
+
+      /**
+       * Returns an iterator optimized for traversing range elements, since it's the most frequent
+       * operation for which ranges are used.
+       */
+      @Override
+      public Iterator<Integer> iterator() {
+        if (step > 0) {
+          return new IncreasingIterator(start, stop, step);
+        } else {
+          return new DecreasingIterator(start, stop, step);
+        }
+      }
+
+      /** @return the start of the range. */
+      public int getStart() {
+        return start;
+      }
+
+      /** @return the stop element (next after the last one) of the range. */
+      public int getStop() {
+        return stop;
+      }
+
+      /** @return the step between each element of the range. */
+      public int getStep() {
+        return step;
+      }
+    }
+
+    private final RangeListView contents;
+
+    private RangeList(int start, int stop, int step) {
+      this.step = step;
+      this.contents = new RangeListView(start, stop, step);
+    }
+
+    @Override
+    public boolean isTuple() {
+      return false;
+    }
+
+    @Override
+    public ImmutableList<Integer> getImmutableList() {
+      return ImmutableList.copyOf(contents);
+    }
+
+    @Override
+    public SkylarkList<Integer> getSlice(
+        Object start, Object end, Object step, Location loc, Mutability mutability)
+        throws EvalException {
+      RangeListView.Slice slice = RangeListView.Slice.from(size(), start, end, step, loc);
+      int substep = slice.step * this.step;
+      int substart = get(slice.start);
+      int substop = get(slice.stop);
```

Comment:
It's really confusing which indices are inclusive and exclusive in the new code. A comment or two might help. The simplest interpretation is that `start` and `stop` are consistently inclusive and exclusive, throughout. In that case, `slice.stop` is not necessarily a valid input to `get()`, when the slice's endpoint is the same as the original range's.

Reply:
Excellent point! :) I've updated code to use [start, stop) convention everywhere